feat(iota): alphanet genesis ceremony

kodemartin opened the pull request:
# Description of change

This patch introduces the necessary changes to perform successfully the genesis-ceremony of the alphanet network with the migrated state and faucet working. The changes can be summed up as follows:

1. Validation of the `token_distribution_schedule` in `Builder::build_and_cache_unsigned_genesis` is performed only after the genesis stake from the migrated state is merged with any cached (vanilla or not) schedule in the `Builder`. This allows to add simple token allocations (e.g. for the faucet) without requiring to satisfy the minimum stake for the validators.
2. The migrated state is now only added through the `add_migration_source` builder method, that accepts `SnapshotSource` value. This populates a new field `migration_sources` that is saved into a dedicated file with calling `Builder::save`. This allows us to reload the migrated state when calling `Builder::load`  to load the builder from disk.
3. `Builder::token_distribution_schedule` has been made more strict, allowing only schedule values that don't contain timelocked stake, in order to avoid conflicts with genesis stake delegated from the migrated state while reloading a saved `Builder`.
4. A few blocks in `Builder::build_and_cache_unsigned_genesis` have been moved into dedicated methodsto separate concerns,  and refactored to accommodate correct reloading of the `Builder`. Other utility functions have been added here and there.

## Links to any relevant issues

The genesis-ceremony has been successfully performed. So this

Resolves #138

kodemartin committed:
feat(iota-swarm-config): introduce NetworkConfigLight

kodemartin committed:
feat(iota{,-test-validator}): pass multiple snapshot sources

kodemartin committed:
fix(iota): restore broken cli-tests

kodemartin committed:
feat(iota): add token allocation in genesis-ceremony

kodemartin committed:
feat(iota): add migrated state in genesis-ceremony

lzpap reviewed (dismissed):
Managed to run a local network with one validator. The node is crashing with some migrated object layout mismatch, but hey, it starts!

lzpap reviewed (commented):
sry, approved the wrong one PR :D

kodemartin committed:
refactor(iota): update genesis.md

kodemartin committed:
feat(iota): override migrated state in genesis-ceremony

This patch allows overriding the migrated state while building
genesis. This is done with the introduction of two optional arguments:

```
Usage: iota genesis-ceremony build-unsigned-checkpoint [OPTIONS]

Options:
      --local-migration-snapshots [<path>...]                  Define paths to local migration snapshots.
      --remote-migration-snapshots [<iota|smr|<full-url>>...]  Remote migration snapshots. [default: iota smr]
  -h, --help                                                   Print help
  -V, --version                                                Print version
```

By default the `iota` and `smr` migration objects are added to genesis.

kodemartin committed:
fix(iota-genesis-builder): validate schedule after stake delegation

kodemartin committed:
fix!(iota): run genesis-ceremony in async runtime

kodemartin committed:
fix(iota-genesis-builder): add migration sources in Builder::load

kodemartin committed:
fix(iota): restore broken ceremony cli test

kodemartin committed:
refactor(iota-genesis-builder): remove debug expression

kodemartin committed:
refactor(iota-genesis-builder): remove duplicate validation of token_distribution_schedule

kodemartin committed:
refactor!(iota-genesis-builder): remove Builder::add_migration_objects

We now rely only on `Builder::add_migration_source`.

kodemartin committed:
refactor(iota-genesis-builder): disallow vanilla token-distribution with timelocked stake

kodemartin committed:
fix(iota-swarm-config): update broken test snapshot

junwei0117 reviewed (approved):
The network is working well with commit [18c4523](https://github.com/iotaledger/iota/commit/18c452375903ff36c751f66708b26c21c26ac902). All nodes, the indexer, and the faucet are healthy and syncing.

kodemartin committed:
fix(iota): restore broken cli-tests

kodemartin committed:
refactor(iota): fix spurious clippy errors from rebase
